ReactiveCocoaTests/RACKVOProxySpec.m

```diff
@@ -0,0 +1,221 @@
+			__block int observedValue1 = 0;
+			__block int observedValue2 = 0;
+
+			RACDisposable *disposable1 = [RACObserve(testObject, testInt) subscribeNext:^(NSNumber *wrappedInt) {
+				observedValue1 = wrappedInt.intValue;
+			}];
+
+			[RACObserve(testObject, testInt) subscribeNext:^(NSNumber *wrappedInt) {
+				observedValue2 = wrappedInt.intValue;
+			}];
+
+			testObject.testInt = 2;
+
+			expect(@(observedValue1)).toEventually(equal(@2));
+			expect(@(observedValue2)).toEventually(equal(@2));
+
+			[disposable1 dispose];
+			testObject.testInt = 3;
+
+			expect(@(observedValue2)).toEventually(equal(@3));
+			expect(@(observedValue1)).to(equal(@2));
+		});
+	});
+
+	qck_describe(@"async", ^{
+		qck_it(@"should handle changes being made on another queue", ^{
+			__block int observedValue = 0;
+			[[[RACObserve(testObject, testInt)
+				skip:1]
+				take:1]
+				subscribeNext:^(NSNumber *wrappedInt) {
+					observedValue = wrappedInt.intValue;
+				}];
+
+			dispatch_async(concurrentQueue, ^{
+				testObject.testInt = 2;
+			});
+
+			dispatch_barrier_sync(concurrentQueue, ^{});
+			expect(@(observedValue)).toEventually(equal(@2));
+		});
+
+		qck_it(@"should handle changes being made on another queue using deliverOn", ^{
+			__block int observedValue = 0;
+			[[[[RACObserve(testObject, testInt)
+				skip:1]
+				take:1]
+				deliverOn:[RACScheduler mainThreadScheduler]]
+				subscribeNext:^(NSNumber *wrappedInt) {
+					observedValue = wrappedInt.intValue;
+				}];
+
+			dispatch_async(concurrentQueue, ^{
+				testObject.testInt = 2;
+			});
+
+			dispatch_barrier_sync(concurrentQueue, ^{});
+			expect(@(observedValue)).toEventually(equal(@2));
+		});
+
+		qck_it(@"async disposal of target", ^{
+			__block int observedValue;
+			[[RACObserve(testObject, testInt)
+				deliverOn:RACScheduler.mainThreadScheduler]
+				subscribeNext:^(NSNumber *wrappedInt) {
+					observedValue = wrappedInt.intValue;
+				}];
+
+			dispatch_async(concurrentQueue, ^{
+				testObject.testInt = 2;
+				testObject = nil;
+			});
+
+			dispatch_barrier_sync(concurrentQueue, ^{});
+			expect(@(observedValue)).toEventually(equal(@2));
+		});
+	});
+
+	qck_describe(@"stress", ^{
+		static const size_t numIterations = 5000;
+
+		__block dispatch_queue_t iterationQueue;
+
+		beforeEach(^{
+			iterationQueue = dispatch_queue_create("org.reactivecocoa.ReactiveCocoa.RACKVOProxySpec.iterationQueue", DISPATCH_QUEUE_CONCURRENT);
+		});
+
+		// ReactiveCocoa/ReactiveCocoa#1122
+		qck_it(@"async disposal of observer", ^{
+			RACSerialDisposable *disposable = [[RACSerialDisposable alloc] init];
+
+			dispatch_apply(numIterations, iterationQueue, ^(size_t index) {
+				RACDisposable *newDisposable = [RACObserve(testObject, testInt) subscribeCompleted:^{}];
+				[[disposable swapInDisposable:newDisposable] dispose];
+
+				dispatch_async(concurrentQueue, ^{
+					testObject.testInt = (int)index;
+				});
+			});
+
+			dispatch_barrier_sync(iterationQueue, ^{
```

Comment:
Ah, I misunderstood `dispatch_apply`. Nonetheless, this code is correct—this is the queue I intended to dispatch to, and it has the same semantics regardless of whether it's already empty or not.

`concurrentQueue` is properly drained at the top, before the test officially completes.